kafka-client/src/test/java/dev/responsive/kafka/internal/db/CassandraFactTableIntegrationTest.java

```diff
@@ -184,4 +187,220 @@ public void shouldRespectSemanticDefaultOnlyTtlForLookups() throws Exception {
     assertThat(expired, Matchers.nullValue());
   }
 
+  @Test
+  public void shouldRespectSemanticKeyBasedTtl() throws Exception {
+    // Given:
+    final var defaultTtl = Duration.ofMillis(30);
+    final Function<String, Optional<TtlDuration>> ttlForKey = k -> {
+      if (k.equals("KEEP_FOREVER")) {
+        return Optional.of(TtlDuration.noTtl());
+      } else if (k.endsWith("DEFAULT_TTL")) {
+        return Optional.empty();
+      } else {
+        final long ttlMilliseconds = Long.parseLong(k);
+        return Optional.of(TtlDuration.of(Duration.ofMillis(ttlMilliseconds)));
+      }
+    };
+    final TtlProvider<String, String> ttlProvider = TtlProvider
+        .<String, String>withDefault(defaultTtl)
+        .fromKey(ttlForKey, Serdes.String());
+    params = ResponsiveKeyValueParams.fact(storeName).withTtlProvider(ttlProvider);
+
+    final var table = client.factFactory().create(RemoteTableSpecFactory.fromKVParams(
+        params,
+        defaultPartitioner(),
+        Optional.of(new TtlResolver<>(false, "changelog-ignored", ttlProvider))
+    ));
+
+    table.init(1);
+
+    final Bytes noTtlKey = serializedKey("KEEP_FOREVER");
+    final Bytes defaultTtlKey = serializedKey("DEFAULT_TTL"); // default is 30ms
+    final Bytes tenMsTtlKey = serializedKey(String.valueOf(10L));
+    final Bytes fiftyMsTtlKey = serializedKey(String.valueOf(50L));
+
+    final byte[] val = new byte[]{1};
+
+    // When:
+    final long insertTimeMs = 0L;
+    client.execute(table.insert(1, noTtlKey, val, insertTimeMs));
+    client.execute(table.insert(1, defaultTtlKey, val, insertTimeMs));
+    client.execute(table.insert(1, tenMsTtlKey, val, insertTimeMs));
+    client.execute(table.insert(1, fiftyMsTtlKey, val, insertTimeMs));
+
+    // Then
+    long lookupTime = Duration.ofMillis(11).toMillis();
```

Comment:
hmm, now that I'm thinking about it we may want to use two orders of magnitude larger TTLs than ~10ms because (a) Cassandra only has TTL on order of seconds, so these would all translate to 0 seconds IIRC and (b) we don't want the TTl to pass and the row to be physically expired before the lookup happens.

Reply:
mm good call

Reply:
@agavra since the build already finished/passed on this PR I'm going to go ahead and merge it as-is, but will address this in the next PR if that's ok